### ISSUE TITLE
[Request] NBSP Localization Fix

### ISSUE THREAD

**Issue by Elvarion** [MinecraftForge/MinecraftForge#3521]
Using Swedish (And French, possibly other languages?) as localization for you OS will cause mods that use numbers above 1000 to be faulty displayed. Instead of a space or a comma it will display the Non-breaking space "NBSP".

Example of NBSP
![](https://i.gyazo.com/7cfe5975d03877a084cbf7348cb768cd.png)

And how it should look (Well look on English(UK) localization.)
I do not really care if its a space or a comma that is being displayed.
As long as the NBSP is not there i am happy
![](https://i.gyazo.com/3d5d1ae824d647be8763eacf435da538.png)

EnderIO made a workaround when they fixed it for the French localization. This fixed it for Swedish also. After contacting TinkersConstruct they also fixed it for their mod. But there is still mods out there that has this issue.

Perhaps we could get a fix for this in Forge so mods do not have to create workarounds for this?

---
Minecraft themselves will most likely not fix this since its not related to Vanilla and really only happens when playing with mods installed.
https://bugs.mojang.com/browse/MC-2595

[Talking to EnderIO about the NBSP](https://github.com/SleepyTrousers/EnderIO/issues/3869)
[Code](https://github.com/SleepyTrousers/EnderIO/blob/040dae0dd1b456758a21240e8186a8783156c407/src/main/java/crazypants/enderio/power/PowerDisplayUtil.java)

[Talking to TinkersConstruct](https://github.com/SlimeKnights/TinkersConstruct/issues/2608)
[Code](https://github.com/SlimeKnights/TinkersConstruct/blob/c381e8c38c66863cd6fc0b28e7652bd2c9ffaa0b/src/main/java/slimeknights/tconstruct/library/Util.java#L31)

**Comment by bs2609** [comment]
Any relation to [MC-88484](https://bugs.mojang.com/browse/MC-88484)?

**Comment by mezz** [comment]
It looks like there are two issues here:
 1. NBSP character should just look like a space. The character sheet in the default resource pack is incorrect.
 2. Minecraft uses the ROOT locale for number formatting instead of the language set in the game

The proper solution for the first issue is fixing the actual font asset or adding a resource pack that changes the NBSP to a blank space. You can take a look in the minecraft jar at `assets/minecraft/textures/font/unicode_page_00.png` which contains the NBSP character. Forge is not in a good position to change this, Mojang is though. 
It may be possible for Forge to re-route all lookups for the NBSP character to the regular space character, that may be the hacky solution to this issue until Mojang fixes it.

The second issue requires changing all formatting functions to be called with the correct Locale. While we could just do a single call to `Locale.setDefault` instead, there would certainly be unintended consequences. There are over 100 calls to `String.format` in vanilla Minecraft without specifying the locale, so this is a big change for a small gain (sorry). This part of the solution must come from Mojang.
For mods, they can properly fix their string formatting by calling `String.format` with a locale.

**Comment by Elvarion** [comment]
Thanks for the reply mezz and shedding some light on this matter

> The proper solution for the first issue is fixing the actual font asset or adding a resource pack that changes the NBSP to a blank space.

Been there, done that. Sure it "does" solve it, however it does look somewhat strange since making the NBSP just blank will create a huge gap between the numbers.
![](https://i.gyazo.com/8a6d14b5ef1ef96d7ad13e0604bf25a3.png)
As to what it should look like
![](https://i.gyazo.com/891028e1de4a6c3e1fd435970d5d577c.png)

> There are over 100 calls to String.format in vanilla Minecraft without specifying the locale, so this is a big change for a small gain (sorry).

Dont be sorry :smile: , I have been living with this since mods came out pretty much. Came back to Minecraft a while ago and noticed that nearly all the NBSP was gone from EnderIO. So checked with other mod devs and asked if  they could fix theirs also. So I thought that maybe Forge could apply a fix for this if it was a easy thing to do, turns out it wasn't. But it never hurts to ask!

Ill just go bug Mojang and mod devs to fix their code :smiley:

**Comment by mezz** [comment]
I will work on the NBSP issue since Mojang closed https://bugs.mojang.com/browse/MC-2595 as a "modding" issue.
I recommend opening a new bug that shows the problem using *only* vanilla and never mention modding. At worst, they will close it as a duplicate and re-open the original issue.

**Comment by mezz** [comment]
I'm not sure I can find a place where vanilla uses this separator, actually. It may be legitimate that they keep the issue closed.

**Comment by mezz** [comment]
Ok found a place, they use thousands separators for some of the game statistics, so you can report that.
Strangely it just uses a comma, and not the space for French. They force their number formatter there to use en_US. d'oh

**Comment by mezz** [comment]
Made a PR here #3522

**Comment by mezz** [comment]
I found the relevant Minecraft issue that may lead to the resolution of this problem. 
[I left a comment explaining how to reproduce the problem here, MC-33398.](https://bugs.mojang.com/browse/MC-33398?focusedCommentId=348586&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-348586)

**Comment by Elvarion** [comment]
All this code talk is not making much sense for me, But im very thankful for everything you guys are doing to help solve this :smiley:

**Comment by williewillus** [comment]
This is fixed in the next 1.11 snapshot so maybe we can backport that or just wait for it?

**Comment by mezz** [comment]
The snapshot seems to just use `Locale.US` and ignores both the system/JVM locale and the game Language. 
I updated the [MC-33398 issue](https://bugs.mojang.com/browse/MC-33398?focusedCommentId=348698&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-348698) with what I found.

From the decompiled jar:
```java
k = NumberFormat.getIntegerInstance(Locale.US);
...
l = new DecimalFormat("########0.00");
```
I'm pretty sure it doesn't matter what locale they use if the decimal format is explicit about using a decimal point like that.
